app/lib/package/models.dart

```diff
@@ -235,12 +236,46 @@ class Package extends db.ExpandoModel<String> {
     uploaders!.remove(userId);
   }
 
+  bool updateLatestVersionReferences(
+    List<PackageVersion> versions, {
+    required Version dartSdkVersion,
+  }) {
+    final oldStableVersion = latestSemanticVersion;
+    final oldPrereleaseVersion = latestPrereleaseSemanticVersion;
+    final oldPreviewVersion = latestPreviewSemanticVersion;
+
+    // reset field values
+    latestVersionKey = null;
+    latestPrereleaseVersionKey = null;
+    latestPreviewVersionKey = null;
+
+    versions.forEach((v) => updateVersion(v,
+        dartSdkVersion: dartSdkVersion, includeRetracted: false));
+
+    if (latestVersionKey == null) {
+      // All versions are retracted, we use the latest regardless of retracted status.
+      versions.forEach((v) => updateVersion(v,
+          dartSdkVersion: dartSdkVersion, includeRetracted: true));
+    }
+
+    final unchanged = oldStableVersion == latestSemanticVersion &&
+        oldPrereleaseVersion == latestPrereleaseSemanticVersion &&
+        oldPreviewVersion == latestPreviewSemanticVersion;
+    if (unchanged) {
+      return false;
+    }
+    updated = DateTime.now().toUtc();
+    return true;
+  }
+
   /// Updates latest stable, prerelease and preview versions and published
   /// timestamp with the new version.
   void updateVersion(
     PackageVersion pv, {
     required Version dartSdkVersion,
+    bool includeRetracted = false,
   }) {
+    if (!includeRetracted && pv.isRetracted) return;
```

Comment:
I'm wondering if this is needed here, or maybe at line 252? (see above)

Reply:
Sure I can move that logic one level up, it's much nicer.